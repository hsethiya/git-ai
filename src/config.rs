use std::collections::HashSet;
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::OnceLock;

use serde::Deserialize;

use crate::git::repository::Repository;

/// Centralized configuration for the application
pub struct Config {
    git_path: String,
    ignore_prompts: bool,
    allow_repositories: HashSet<String>,
}
#[derive(Deserialize)]
struct FileConfig {
    #[serde(default)]
    git_path: Option<String>,
    #[serde(default)]
    ignore_prompts: Option<bool>,
    #[serde(default)]
    allow_repositories: Option<Vec<String>>,
}

static CONFIG: OnceLock<Config> = OnceLock::new();

impl Config {
    /// Initialize the global configuration exactly once.
    /// Safe to call multiple times; subsequent calls are no-ops.
    #[allow(dead_code)]
    pub fn init() {
        let _ = CONFIG.get_or_init(|| build_config());
    }

    /// Access the global configuration. Lazily initializes if not already initialized.
    pub fn get() -> &'static Config {
        CONFIG.get_or_init(|| build_config())
    }

    /// Returns the command to invoke git.
    pub fn git_cmd(&self) -> &str {
        &self.git_path
    }

    pub fn get_ignore_prompts(&self) -> bool {
        self.ignore_prompts
    }

    pub fn is_allowed_repository(&self, repository: &Option<Repository>) -> bool {
        // If allowlist is empty, allow everything
        if self.allow_repositories.is_empty() {
            return true;
        }

        // If allowlist is defined, only allow repos whose remotes match the list
        if let Some(repository) = repository {
            match repository.remotes_with_urls().ok() {
                Some(remotes) => remotes
                    .iter()
                    .any(|remote| self.allow_repositories.contains(&remote.1)),
                None => false, // Can't verify, deny by default when allowlist is active
            }
        } else {
            false // No repository provided, deny by default when allowlist is active
        }
    }

    /// Returns whether prompts should be ignored (currently unused by internal APIs).
    #[allow(dead_code)]
    pub fn ignore_prompts(&self) -> bool {
        self.ignore_prompts
    }
}

fn build_config() -> Config {
    let file_cfg = load_file_config();
    let ignore_prompts = file_cfg
        .as_ref()
        .and_then(|c| c.ignore_prompts)
        .unwrap_or(false);
    let allow_repositories = file_cfg
        .as_ref()
        .and_then(|c| c.allow_repositories.clone())
        .unwrap_or(vec![])
        .into_iter()
        .collect();

    let git_path = resolve_git_path(&file_cfg);

    Config {
        git_path,
        ignore_prompts,
        allow_repositories,
    }
}

fn resolve_git_path(file_cfg: &Option<FileConfig>) -> String {
    // 1) From config file
    if let Some(cfg) = file_cfg {
        if let Some(path) = cfg.git_path.as_ref() {
            let trimmed = path.trim();
            if !trimmed.is_empty() {
                let p = Path::new(trimmed);
                if is_executable(p) {
                    return trimmed.to_string();
                }
            }
        }
    }

    // 2) Probe common locations across platforms
    let candidates: &[&str] = &[
        // macOS Homebrew (ARM and Intel)
        "/opt/homebrew/bin/git",
        "/usr/local/bin/git",
        // Common Unix paths
        "/usr/bin/git",
        "/bin/git",
        "/usr/local/sbin/git",
        "/usr/sbin/git",
        // Windows Git for Windows
        r"C:\\Program Files\\Git\\bin\\git.exe",
        r"C:\\Program Files (x86)\\Git\\bin\\git.exe",
    ];

    if let Some(found) = candidates.iter().map(Path::new).find(|p| is_executable(p)) {
        return found.to_string_lossy().to_string();
    }

    // 3) Fatal error: no real git found
    eprintln!(
        "Fatal: Could not locate a real 'git' binary.\n\
         Expected a valid 'git_path' in {cfg_path} or in standard locations.\n\
         Please install Git or update your config JSON.",
        cfg_path = config_file_path()
            .map(|p| p.to_string_lossy().to_string())
            .unwrap_or_else(|| "~/.git-ai/config.json".to_string()),
    );
    std::process::exit(1);
}

fn load_file_config() -> Option<FileConfig> {
    let path = config_file_path()?;
    let data = fs::read(&path).ok()?;
    serde_json::from_slice::<FileConfig>(&data).ok()
}

fn config_file_path() -> Option<PathBuf> {
    #[cfg(windows)]
    {
        let home = env::var("USERPROFILE").ok()?;
        Some(Path::new(&home).join(".git-ai").join("config.json"))
    }
    #[cfg(not(windows))]
    {
        let home = env::var("HOME").ok()?;
        Some(Path::new(&home).join(".git-ai").join("config.json"))
    }
}

fn is_executable(path: &Path) -> bool {
    if !path.exists() || !path.is_file() {
        return false;
    }
    // Basic check: existence is sufficient for our purposes; OS will enforce exec perms.
    // On Unix we could check permissions, but many filesystems differ. Keep it simple.
    true
}
