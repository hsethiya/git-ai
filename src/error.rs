use std::fmt;

#[derive(Debug)]
pub enum GitAiError {
    GitError(git2::Error),
    IoError(std::io::Error),
    /// Errors from invoking the git CLI that exited with a non-zero status
    GitCliError {
        code: Option<i32>,
        stderr: String,
        args: Vec<String>,
    },
    JsonError(serde_json::Error),
    Utf8Error(std::str::Utf8Error),
    FromUtf8Error(std::string::FromUtf8Error),
    PresetError(String),
    Generic(String),
}

impl fmt::Display for GitAiError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            GitAiError::GitError(e) => write!(f, "Git error: {}", e),
            GitAiError::IoError(e) => write!(f, "IO error: {}", e),
            GitAiError::GitCliError { code, stderr, args } => match code {
                Some(c) => write!(f, "Git CLI ({}) failed with exit code {}: {}", args.join(" "), c, stderr),
                None => write!(f, "Git CLI ({}) failed: {}", args.join(" "), stderr),
            },
            GitAiError::JsonError(e) => write!(f, "JSON error: {}", e),
            GitAiError::Utf8Error(e) => write!(f, "UTF-8 error: {}", e),
            GitAiError::FromUtf8Error(e) => write!(f, "From UTF-8 error: {}", e),
            GitAiError::PresetError(e) => write!(f, "{}", e),
            GitAiError::Generic(e) => write!(f, "Generic error: {}", e),
        }
    }
}

impl std::error::Error for GitAiError {}

impl From<git2::Error> for GitAiError {
    fn from(err: git2::Error) -> Self {
        GitAiError::GitError(err)
    }
}

impl From<std::io::Error> for GitAiError {
    fn from(err: std::io::Error) -> Self {
        GitAiError::IoError(err)
    }
}

impl From<serde_json::Error> for GitAiError {
    fn from(err: serde_json::Error) -> Self {
        GitAiError::JsonError(err)
    }
}

impl From<std::str::Utf8Error> for GitAiError {
    fn from(err: std::str::Utf8Error) -> Self {
        GitAiError::Utf8Error(err)
    }
}

impl From<std::string::FromUtf8Error> for GitAiError {
    fn from(err: std::string::FromUtf8Error) -> Self {
        GitAiError::FromUtf8Error(err)
    }
}
