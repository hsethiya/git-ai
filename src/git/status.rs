use crate::config::Config;
use crate::error::GitAiError;
use git2::Repository;
use std::process::Command;
use std::str;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StatusCode {
    Unmodified,
    Modified,
    Added,
    Deleted,
    Renamed,
    Copied,
    Unmerged,
    Untracked,
    Ignored,
    Unknown(char),
}

impl From<char> for StatusCode {
    fn from(value: char) -> Self {
        match value {
            '.' => StatusCode::Unmodified,
            'M' => StatusCode::Modified,
            'A' => StatusCode::Added,
            'D' => StatusCode::Deleted,
            'R' => StatusCode::Renamed,
            'C' => StatusCode::Copied,
            'U' => StatusCode::Unmerged,
            '?' => StatusCode::Untracked,
            '!' => StatusCode::Ignored,
            other => StatusCode::Unknown(other),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EntryKind {
    Ordinary,
    Rename,
    Copy,
    Unmerged,
    Untracked,
    Ignored,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StatusEntry {
    pub path: String,
    pub staged: StatusCode,
    pub unstaged: StatusCode,
    pub kind: EntryKind,
    pub orig_path: Option<String>,
}

pub fn status_porcelainv2(repo: &Repository) -> Result<Vec<StatusEntry>, GitAiError> {
    let workdir = repo
        .workdir()
        .ok_or_else(|| GitAiError::Generic("Repository has no working directory".into()))?;

    let output = Command::new(Config::get().git_cmd())
        .arg("status")
        .arg("--porcelain=v2")
        .arg("-z")
        .current_dir(workdir)
        .output()?;

    if !output.status.success() {
        return Err(GitAiError::Generic(format!(
            "git status exited with status {}",
            output.status
        )));
    }

    parse_porcelain_v2(&output.stdout)
}

fn parse_porcelain_v2(data: &[u8]) -> Result<Vec<StatusEntry>, GitAiError> {
    let mut entries = Vec::new();
    let mut parts = data
        .split(|byte| *byte == 0)
        .filter(|slice| !slice.is_empty())
        .peekable();

    while let Some(raw) = parts.next() {
        let record = str::from_utf8(raw)?;
        let mut chars = record.chars();
        let tag = chars
            .next()
            .ok_or_else(|| GitAiError::Generic("Unexpected empty porcelain v2 record".into()))?;

        match tag {
            '1' | 'u' => {
                let mut fields = record.splitn(9, ' ');
                let _ = fields.next(); // tag
                let xy = fields
                    .next()
                    .ok_or_else(|| GitAiError::Generic("Missing XY field".into()))?;
                if xy.len() != 2 {
                    return Err(GitAiError::Generic(format!(
                        "Unexpected XY field length: {}",
                        xy
                    )));
                }
                let staged = StatusCode::from(xy.chars().next().unwrap());
                let unstaged = StatusCode::from(xy.chars().nth(1).unwrap());

                // skip submodule/metadata fields to capture path
                for _ in 0..6 {
                    fields.next();
                }

                let path = fields
                    .next()
                    .ok_or_else(|| GitAiError::Generic("Missing path field".into()))?
                    .to_string();

                entries.push(StatusEntry {
                    path,
                    staged,
                    unstaged,
                    kind: if matches!(staged, StatusCode::Unmerged)
                        || matches!(unstaged, StatusCode::Unmerged)
                    {
                        EntryKind::Unmerged
                    } else {
                        EntryKind::Ordinary
                    },
                    orig_path: None,
                });
            }
            '2' => {
                let mut fields = record.splitn(10, ' ');
                let _ = fields.next(); // tag
                let xy = fields
                    .next()
                    .ok_or_else(|| GitAiError::Generic("Missing XY field".into()))?;
                if xy.len() != 2 {
                    return Err(GitAiError::Generic(format!(
                        "Unexpected XY field length: {}",
                        xy
                    )));
                }
                let staged = StatusCode::from(xy.chars().next().unwrap());
                let unstaged = StatusCode::from(xy.chars().nth(1).unwrap());

                // skip submodule/metadata fields
                for _ in 0..7 {
                    fields.next();
                }

                let path = fields
                    .next()
                    .ok_or_else(|| GitAiError::Generic("Missing path field".into()))?
                    .to_string();

                let orig_path_bytes = parts.next().ok_or_else(|| {
                    GitAiError::Generic("Missing original path for rename/copy".into())
                })?;
                let orig_path = str::from_utf8(orig_path_bytes)?.to_string();

                let kind = match staged {
                    StatusCode::Renamed => EntryKind::Rename,
                    StatusCode::Copied => EntryKind::Copy,
                    _ => EntryKind::Ordinary,
                };

                entries.push(StatusEntry {
                    path,
                    staged,
                    unstaged,
                    kind,
                    orig_path: Some(orig_path),
                });
            }
            '?' => {
                let path = record.strip_prefix("? ").unwrap_or(record).to_string();

                entries.push(StatusEntry {
                    path,
                    staged: StatusCode::Unmodified,
                    unstaged: StatusCode::Untracked,
                    kind: EntryKind::Untracked,
                    orig_path: None,
                });
            }
            '!' => {
                let path = record.strip_prefix("! ").unwrap_or(record).to_string();

                entries.push(StatusEntry {
                    path,
                    staged: StatusCode::Unmodified,
                    unstaged: StatusCode::Ignored,
                    kind: EntryKind::Ignored,
                    orig_path: None,
                });
            }
            other => {
                return Err(GitAiError::Generic(format!(
                    "Unsupported porcelain v2 record tag: {}",
                    other
                )));
            }
        }
    }

    Ok(entries)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn parse_varied_porcelain_v2_records() {
        // Construct a blob of porcelain v2 entries covering tracked, renamed, copied,
        // unmerged, untracked, and ignored states with spaces and special characters.
        let mut raw = Vec::new();
        raw.extend_from_slice(b"1 MM N... 100644 100644 100644 1111111111111111111111111111111111111111 2222222222222222222222222222222222222222 src/lib.rs\0");
        raw.extend_from_slice(b"1 AM N... 100644 100755 100755 3333333333333333333333333333333333333333 4444444444444444444444444444444444444444 src/bin/cli.rs\0");
        raw.extend_from_slice(b"1 .U N... 100644 100644 100644 5555555555555555555555555555555555555555 6666666666666666666666666666666666666666 src/conflict.rs\0");
        raw.extend_from_slice(b"2 R. N... 100644 100644 100644 7777777777777777777777777777777777777777 8888888888888888888888888888888888888888 80 src/utils/helpers.rs\0old utils/helpers.rs\0");
        raw.extend_from_slice(b"2 C. N... 100644 100644 100644 9999999999999999999999999999999999999999 0000000000000000000000000000000000000000 60 scripts/setup.sh\0scripts/setup-old.sh\0");
        raw.extend_from_slice(b"1 D. N... 100644 000000 000000 aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa 0000000000000000000000000000000000000000 docs/README.md\0");
        raw.extend_from_slice(b"1 A. N... 000000 100644 100644 0000000000000000000000000000000000000000 bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb \"space dir\"/new file.txt\0");
        raw.extend_from_slice(b"1 M. N... 100644 100644 100644 cccccccccccccccccccccccccccccccccccccccc dddddddddddddddddddddddddddddddddddddddd path/with->symbol.rs\0");
        raw.extend_from_slice(b"? assets/logo (1).svg\0");
        raw.extend_from_slice(b"? dir with spaces/file name [draft].md\0");
        raw.extend_from_slice(b"! target/.keep\0");
        raw.extend_from_slice(b"u UU N... 100644 100644 100644 eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee ffffffffffffffffffffffffffffffffffffffff 1 2 3 some unmerged/path.txt\0");

        let entries: Vec<StatusEntry> = parse_porcelain_v2(&raw).expect("parse succeeds");

        // High-level assertions about the parsed content
        assert_eq!(entries.len(), 12);
        assert!(
            entries
                .iter()
                .any(|e| e.path == "src/lib.rs" && e.staged == StatusCode::Modified)
        );
        assert!(entries.iter().any(|e| e.kind == EntryKind::Rename
            && e.orig_path.as_deref() == Some("old utils/helpers.rs")));
        assert!(
            entries.iter().any(|e| e.kind == EntryKind::Copy
                && e.orig_path.as_deref() == Some("scripts/setup-old.sh"))
        );
        assert!(entries.iter().any(|e| e.kind == EntryKind::Unmerged));
        assert!(
            entries
                .iter()
                .any(|e| matches!(e.unstaged, StatusCode::Untracked))
        );
        assert!(
            entries
                .iter()
                .any(|e| matches!(e.unstaged, StatusCode::Ignored))
        );

        assert_debug_snapshot!(entries);
    }
}
