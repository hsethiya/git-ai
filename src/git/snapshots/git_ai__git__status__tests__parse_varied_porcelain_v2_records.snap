---
source: src/git/status.rs
expression: entries
---
[
    StatusEntry {
        path: "src/lib.rs",
        staged: Modified,
        unstaged: Modified,
        kind: Ordinary,
        orig_path: None,
    },
    StatusEntry {
        path: "src/bin/cli.rs",
        staged: Added,
        unstaged: Modified,
        kind: Ordinary,
        orig_path: None,
    },
    StatusEntry {
        path: "src/conflict.rs",
        staged: Unmodified,
        unstaged: Unmerged,
        kind: Unmerged,
        orig_path: None,
    },
    StatusEntry {
        path: "src/utils/helpers.rs",
        staged: Renamed,
        unstaged: Unmodified,
        kind: Rename,
        orig_path: Some(
            "old utils/helpers.rs",
        ),
    },
    StatusEntry {
        path: "scripts/setup.sh",
        staged: Copied,
        unstaged: Unmodified,
        kind: Copy,
        orig_path: Some(
            "scripts/setup-old.sh",
        ),
    },
    StatusEntry {
        path: "docs/README.md",
        staged: Deleted,
        unstaged: Unmodified,
        kind: Ordinary,
        orig_path: None,
    },
    StatusEntry {
        path: "\"space dir\"/new file.txt",
        staged: Added,
        unstaged: Unmodified,
        kind: Ordinary,
        orig_path: None,
    },
    StatusEntry {
        path: "path/with->symbol.rs",
        staged: Modified,
        unstaged: Unmodified,
        kind: Ordinary,
        orig_path: None,
    },
    StatusEntry {
        path: "assets/logo (1).svg",
        staged: Unmodified,
        unstaged: Untracked,
        kind: Untracked,
        orig_path: None,
    },
    StatusEntry {
        path: "dir with spaces/file name [draft].md",
        staged: Unmodified,
        unstaged: Untracked,
        kind: Untracked,
        orig_path: None,
    },
    StatusEntry {
        path: "target/.keep",
        staged: Unmodified,
        unstaged: Ignored,
        kind: Ignored,
        orig_path: None,
    },
    StatusEntry {
        path: "1 2 3 some unmerged/path.txt",
        staged: Unmerged,
        unstaged: Unmerged,
        kind: Unmerged,
        orig_path: None,
    },
]
